extensions/ql-vscode/src/telemetry.ts

```diff
@@ -0,0 +1,159 @@
+      client.addTelemetryProcessor((envelope) => {
+        if (LOG_TELEMETRY.getValue<boolean>()) {
+          logger.log(`Telemetry: ${JSON.stringify(envelope)}`);
+        }
+        return true;
+      });
+    }
+
+    ctx.subscriptions.push(reporter);
+  }
+}
+
+export function sendCommandUsage(name: string, executionTime: number, error?: Error) {
+  if (!reporter) {
+    return;
+  }
+  const status = !error
+    ? CommandCompletion.Success
+    : error instanceof UserCancellationException
+      ? CommandCompletion.Cancelled
+      : CommandCompletion.Failed;
+
+  const isCanary = (!!CANARY_FEATURES.getValue<boolean>()).toString();
+
+  reporter.sendTelemetryEvent(
+    'command-usage',
+    {
+      name,
+      status,
+      isCanary
+    },
+    { executionTime }
+  );
+
+  // if this is a true error, also report it
+  if (status === CommandCompletion.Failed && error) {
+    const redactedError = {
+      name: error.name,
+      stack: error.stack,
+      message: '<MESSAGE REDACTED>'
+    };
+    reporter.sendTelemetryException(
+      redactedError,
+      {
+        type: 'command-usage',
+        name,
+        status,
+        isCanary
+      },
+      { executionTime }
+    );
+  }
+}
+
+function registerListener(extension: Extension<any>, ctx: ExtensionContext) {
+  if (!listener) {
+    listener = workspace.onDidChangeConfiguration(async e => {
+      if (e.affectsConfiguration('codeQL.telemetry.enableTelemetry')) {
+        await initializeTelemetry(extension, ctx);
+      }
+
+      // TODO: re-request telemetry so that users can see the dialog again.
+      // Re-request if codeQL.canary is being set to `true` and telemetry
+      // is not currently enabled.
+      if (
+        e.affectsConfiguration('codeQL.canary') &&
+        CANARY_FEATURES.getValue() &&
+        !ENABLE_TELEMETRY.getValue()
+      ) {
+        await ctx.globalState.update('telemetry-request-viewed', false);
+        await requestTelemetryPermission(ctx);
+      }
+    });
+    ctx.subscriptions.push(listener);
+  }
+}
+
+async function requestTelemetryPermission(ctx: ExtensionContext) {
+  if (!ctx.globalState.get('telemetry-request-viewed')) {
+    // if global telemetry is disabled, avoid showing the dialog or making any changes
+    let result = undefined;
+    if (workspace.getConfiguration().get<boolean>('telemetry.enableTelemetry')) {
+      // Extension won't start until this completes. So, set a timeout here in order
+      // to ensure the extension continues even if the user doesn't make a choice.
+      result = await Promise.race([
+        showBinaryChoiceDialog(
+          'Do we have your permission to collect anonymous usage statistics help us improve CodeQL for VSCode? See [TELEMETRY.md](https://github.com/github/vscode-codeql/blob/93831e28597bb08567db9844737efd9ff188fc3a/.github/TELEMETRY.md)',
+          { modal: false }
+        ),
+        new Promise(resolve => setTimeout(resolve, 10000))
```

Comment:
Is this timeout enough for a reader to take in the message and decide? What if they decide to follow the link and read?

Shame that the extension won't start if this isn't resolved. What do others to? Is it possible to have it continue with initialisation (without telemetry)?

Reply:
It's possible that the 10s timeout is not long enough. But if they miss it the first time, they will be notified again the next time they restart.

The salesforce extension uses a modal dialog. Doing the same would make this whole section much simpler, though it would be a worse experience for the user. I'm a little torn.

Perhaps we could try out a modal dialog with internal users first and see how they react. It would be easy to switch back to non-modal.